src/bin/ion/commands/dump.rs

```diff
@@ -216,3 +236,59 @@ fn transcribe_n_values<W: IonWriter>(
     writer.flush()?;
     Ok(index)
 }
+
+/// Autodetects a compressed byte stream and wraps the original reader
+/// into a reader that transparently decompresses.
+///
+/// To support non-seekable readers like `Stdin`, we could have used a
+/// full-blown buffering wrapper with unlimited rewinds, but since we only
+/// need the first few magic bytes at offset 0, we cheat and instead make a
+/// `Chain` reader from the buffered header followed by the original reader.
+///
+/// The choice of `Chain` type here is not quite necessary: it could have
+/// been simply `dyn BufRead`, but there is no `ToIonDataSource` trait
+/// implementation for `dyn BufRead` at the moment.
+type AutoDecompressingReader = Chain<Box<dyn BufRead>, Box<dyn BufRead>>;
+
+fn auto_decompressing_reader<R>(
+    mut reader: R,
+    header_len: usize,
+) -> IonResult<AutoDecompressingReader>
+where
+    R: BufRead + 'static,
+{
+    // read header
+    let mut header_bytes = vec![0; header_len];
+    let nread = reader.read(&mut header_bytes)?;
```

Comment:
Does this need a loop to handle the (very unlikely) case where `read` returns a number of bytes that is more than zero but less than `INFER_HEADER_LENGTH`?

Reply:
Yes, let me add that, just to be on the safe side.